src/stats/stats_mempool.h

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_STATS_MEMPOOL_H
+#define BITCOIN_STATS_MEMPOOL_H
+
+#include <amount.h>
+#include "scheduler.h"
+#include <sync.h>
+
+#include <atomic>
+#include <stdlib.h>
+#include <vector>
+
+struct CStatsMempoolSample {
+
+    /* time delta to last item as 16bit unsigned integer
+     * allows a max delta of ~18h */
+    uint16_t timeDelta;
+
+    size_t txCount;     //!<transaction count
+    size_t dynMemUsage; //!<dynamic mempool usage
+    CAmount minFeePerK;  //!<min fee per K
+};
+
+typedef std::vector<struct CStatsMempoolSample> MempoolSamplesVector;
+
+class CStatsMempool
+{
+private:
+
+    /* caches */
+    static std::atomic<size_t> cacheMempoolSize;
+    static std::atomic<size_t> cacheMempoolDynamicMemoryUsage;
+    static std::atomic<CAmount> cacheMempoolMinRelayFee;
+    mutable CCriticalSection cs_mempool_stats; //!<protects the sample/max/time vectors
+
+    std::atomic<uint64_t> startTime; //start time
+    std::vector<MempoolSamplesVector> vSamplesPerPrecision; //!<multiple samples vector per precision level
+    std::vector<size_t> vMaxSamplesPerPrecision; //!<vector with the maximum amount of samples per precision level
+    std::vector<size_t> vTimeLastSample; //!<vector with the maximum amount of samples per precision level
```

Comment:
In commit "Add mempool statistics collector"

All three of these vectors have the same length (number of precision levels). Seems like it would be cleaner to have one vector containing a struct with all the information that needs to be stored for a given precision level. It would also make the c++ data structure more consistent with the json data structure returned from rpc.